src/filesystem.cc

```diff
@@ -1473,22 +1507,46 @@ S3FileSystem::S3FileSystem(
     }
   }
 
-  if ((secret_key != NULL) && (key_id != NULL)) {
-    client_ = s3::S3Client(
+  if (secret_key.size() && key_id.size()) {
+    client_ = new s3::S3Client(
         credentials, config,
         Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never,
         /*useVirtualAdressing*/ false);
 
   } else {
-    client_ = s3::S3Client(
+    client_ = new s3::S3Client(
         config, Aws::Client::AWSAuthV4Signer::PayloadSigningPolicy::Never,
         /*useVirtualAdressing*/ false);
   }
 }
 
 S3FileSystem::~S3FileSystem()
 {
-  Aws::ShutdownAPI(options_);
+  // should delete all client_ before shutdown
+  delete client_;
+  // shutdown aws api if no more S3 instances
+  // FIXME: Calling Aws::ShutdownAPI will result in segmentation fault.
```

Comment:
Bug in AWS client? Is there an issue to track that?

Reply:
The closest one was in 2019 ([link](https://github.com/aws/aws-sdk-cpp/issues/1263)), but was dismissed because of the calling order (`InitAPI`-> do things -> `ShutdownAPI`).

I suspect this is because the `ShutdownAPI` is called when the program is shutting down, and the order might be slightly different from [the required ordering](https://github.com/triton-inference-server/core/pull/103#discussion_r933497563).